File: 322.cpp
/*
 * @lc app=leetcode.cn id=322 lang=cpp
 *
 * [322] 零钱兑换
 */

// @lc code=start
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        int dp[amount+1];
        dp[0] = 0;
        for (int size = 1; size <= amount; ++size) {
            if (size < coins[0]) {
                dp[size] = -1;
            }else if (size == coins[0]) {
                dp[size] = 1;
            }else {
                if (dp[size-coins[0]] > 0) {
                    dp[size] = dp[size-coins[0]] + 1;
                } else {
                    dp[size] = -1;
                }
            }
        }
        for (int item = 1; item < n; ++item) {
            for (int size = 1; size <= amount; ++size) {
                if (size == coins[item]) {
                    dp[size] = 1;
                }else if (size > coins[item]) {
                    if (dp[size-coins[item]] > 0) {
                        if (dp[size] > 0) {
                            dp[size] = min(dp[size], dp[size-coins[item]] + 1);
                        } else {
                            dp[size] = dp[size-coins[item]] + 1;
                        }
                    }
                }
            }
        }
        return dp[amount];
    }
};
// @lc code=end

File: 322.py
#
# @lc app=leetcode.cn id=322 lang=python3
#
# [322] 零钱兑换
#
from typing import List
# @lc code=start
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [0] * (amount + 1)
        coins.sort()
        for i in range(1, amount+1):
            tmp = float('inf')
            for j in coins:
                if j > i:
                    break
                tmp = min(tmp, dp[i-j])
            dp[i] = tmp + 1
        if dp[-1] == float('inf'):
            return -1
        return dp[-1]
# @lc code=end

sol = Solution()
print(sol.coinChange([474,83,404,3], 264))

